### ISSUE TITLE
hover over type with mutual recursion fails silently

### ISSUE THREAD

**Issue by ulugbekna** [ocaml/merlin#1216]
Hover over a type using a mutually recursive type crashes ocaml-lsp silently (ocaml-lsp usually shows thrown exceptions, but not in this case). Say, we have 

```ocaml
type t = s
and s = string
```

hovering over `s` crashes ocaml-lsp with no stack trace, so possibly segfaulting 🤷‍♂️

Note: the problem is discovered using ocaml-lsp, so that might not be merlin's fault; sorry for the noise if that's the case.

**Comment by trefis** [comment]
I don't know what hovering is supposed to trigger in lsp; two reasonable guesses would be "show the type" or "show occurences", both work for me on your example.

**Comment by ulugbekna** [comment]
Thanks for a prompt response!

**Comment by ulugbekna** [comment]
Example I gave in the issue description fails with a stack overflow for me when I run command 

```sh
ocamlmerlin single type-enclosing -position 1:10 -verbosity 1 -\
  > -filename ./test.ml < ./test.ml | jq
```
"hovering over" `s` in 

```ocaml
type t = s and s = string;;
```

@trefis  were you running the same merlin command when you said it was working well for you? 

---

Btw, ocaml-lsp uses `type enclosing` query command for hovering feature.

---

Strangely, for this case, `utop` on `#show t` returns `type nonrec t = t`, which is wrong (?), but `ocaml` returns `type nonrec t = s`. Utop's behavior seems to hint why we could be getting a stack overflow.

**Comment by trefis** [comment]
I had the `verbosity` flag set at 0 (which is the default), but indeed when you increase the value, then you get a stack overflow.
Thanks for the report, we'll look into it.

Btw: is `-verbosity 1` the default in lsp? That seems like it could be a mistake.

**Comment by ulugbekna** [comment]
> Btw: is -verbosity 1 the default in lsp? 

Yes, see this [line](https://github.com/ocaml/ocaml-lsp/blob/cb8aba2238ba5b0e5a6fe0a3180f25077496f75a/ocaml-lsp-server/src/document.ml#L136)

---

> That seems like it could be a mistake.

I don't fully understand what `verbosity` does, but sometimes it can return strange types. Fro example, for a file `t.ml` with contents:

```ocaml
type 'a t = 'a
```

we can run `type-enclosing`  and see a strange type `type 'a t = 'b t` as the head of the list `value`

```sh
❯ ocamlmerlin single type-enclosing -position 1:9 -verbosity 0 -\  
  > -filename ./t.ml < ./t.ml | jq
{
  "class": "return",
  "value": [
    {
      "start": {
        "line": 1,
        "col": 8
      },
      "end": {
        "line": 1,
        "col": 9
      },
      "type": "type 'a t = 'b t",
      "tail": "no"
    },
    {
      "start": {
        "line": 1,
        "col": 0
      },
      "end": {
        "line": 1,
        "col": 14
      },
      "type": "type 'a t = 'a",
      "tail": "no"
    }
  ],
  "notifications": [],
  "timing": {
    "clock": 4,
    "cpu": 4,
    "query": 1,
    "pp": 0,
    "reader": 2,
    "ppx": 0,
    "typer": 2,
    "error": 0
  }
}
```

OCaml-LSP takes the head of this list, i.e., `type 'a t = 'b t`, which is a puzzling type to show a user. This may've been one of the reasons OCaml-LSP uses `verbosity 1`. cc @rgrinberg

**Comment by rgrinberg** [comment]
> Btw: is -verbosity 1 the default in lsp? That seems like it could be a mistake.

Yes, I don't see why it it's 1. I'll revert back to 0.
